ssl/statem/extensions_srvr.c

```diff
@@ -802,8 +803,9 @@ int tls_parse_ctos_cookie(SSL *s, PACKET *pkt, unsigned int context, X509 *x,
     }
 
     /* We tolerate a cookie age of up to 10 minutes (= 60 * 10 seconds) */
-    now = (unsigned long)time(NULL);
-    if (tm > now || (now - tm) > 600) {
+    now = time(NULL);
+    if (OPENSSL_time_t_compare(now, 0, (time_t)tm, 0) < 0
+            || OPENSSL_time_t_compare(now, 0, (time_t)tm, 600) > 0) {
```

Comment:
This can't work because this comparison needs to be done in 32 bit arithmetic.

Reply:
No, it doesn't; it should be done in 64-bit arithmetic, regardless of time_t size.
Can you please explain why 32-bit arithmetic needs to be used?

Reply:
The time stored in the ticket is currently a `uint32_t`.  But we could easily change that...

Reply:
Ok, the original code was likely also problematic, because on a 64-bit system only 4 bytes of the 8-byte `long` were filled in, and thus the 32-bit cookie time value was not sign-extended. Because it's not sign-extended, the cookie time is effectively treated as unsigned thrown into a signed type; treating times pre-1970 as post-2038.

Reply:
> The time stored in the cookie is currently a `uint32_t`. But we could easily change that...

Yeah, it probably should be a `uint64_t`

Reply:
The time is put into the cookie as:
```
WPACKET_put_bytes_u32(pkt, (unsigned int)time(NULL))
```
Past 2038, for a 64-bit `time_t`, when it's extracted out, it will continue to be a treated properly, because it's not sign-extended when it's extracted. This effectively gives OpenSSL's cookie an 68 extra years (until 2106).

Past 2038, for a 32-bit `time_t`, there are problems, as we don't know what the behavior of `time(NULL)` is, it could wrap, it could give -1, it could give MAX_INT.

Reply:
The value in `tm` is always going to be only 32-bits, regardless of `sizeof(tm)`.
Common situations:
If `sizeof(unsigned long) == sizeof(time_t) == 4` then the *system* can only safely handle time until 2038; as the time value put into the cookie is unknown.
If `sizeof(unsigned long) == sizeof(time_t) == 8` then the code can only safely handle time until 2106. The unsigned `tm` is not sign-extended when converted to a `time_t`. The subsequent 64-bit arithmetic works.
Less common situations:
If `sizeof(unsigned long) == 4 && sizeof(time_t) == 8` then the code can only safely handle time until 2106. The unsigned `tm` is not signed extended when converted to a `time_t`. The subsequent 64-bit arithmetic works.
if `sizeof(unsigned long) == 8 && sizeof(time_t) == 4`, then the *system* can only safely handle time until 2038; as the time value put into the cookie is unknown.

Reply:
The `sizeof(unsigned long) == 4 && sizeof(time_t) == 8` is actually common for "recent" 32 bit architecture.

Is there a reason we can't just change tm to an uint64_t?